holidays/countries/vietnam.py

```diff
@@ -29,64 +26,60 @@ class Vietnam(HolidayBase):
+            self._add_new_years_day("International New Year's Day")
+        )
 
         # Vietnamese Kings' Commemoration Day
         # https://en.wikipedia.org/wiki/H%C3%B9ng_Kings%27_Festival
         if year >= 2007:
-            hol_date = self.get_solar_date(year, 3, 10)
-            self[hol_date] = "Hung Kings Commemoration Day"
+            observed_dates.add(
+                self._add_korean_calendar_holiday(
+                    "Hung Kings Commemoration Day", 3, 10
+                )
+            )
 
         # Liberation Day/Reunification Day
-        self[date(year, APR, 30)] = "Liberation Day/Reunification Day"
+        observed_dates.add(
+            self._add_holiday("Liberation Day/Reunification Day", APR, 30)
+        )
 
         # International Labor Day
-        self[date(year, MAY, 1)] = "International Labor Day"
+        observed_dates.add(self._add_labour_day("International Labor Day"))
 
         # Independence Day
-        self[date(year, SEP, 2)] = "Independence Day"
-
+        observed_dates.add(self._add_holiday("Independence Day", SEP, 2))
         if self.observed:
-            for dt in sorted(list(self.keys())):
-                if dt.year == year:
-                    self._add_observed(dt)
+            for dt in sorted(observed_dates):
+                if not self._is_weekend(dt):
+                    continue
+                next_workday = dt + td(days=+1)
+                while (
+                    self._is_weekend(next_workday)
+                    or next_workday in observed_dates
+                ):
+                    next_workday += td(days=+1)
+                observed_dates.add(
+                    self._add_holiday(f"{self[dt]} (Observed)", next_workday)
+                )
 
         # Lunar New Year
-        names = (
-            (-1, "Vietnamese New Year's Eve"),
-            (0, "Vietnamese New Year"),
-            (1, "The second day of Tet Holiday"),
-            (2, "The third day of Tet Holiday"),
-            (3, "The forth day of Tet Holiday"),
-            (4, "The fifth day of Tet Holiday"),
-        )
-        hol_date = self.get_solar_date(year, 1, 1)
-        for d, name in names:
-            self[(hol_date + td(days=+d))] = name
-
-    # convert lunar calendar date to solar
-    def get_solar_date(self, year, month, day):
-        self.korean_cal.setLunarDate(year, month, day, False)
-        return date(
-            self.korean_cal.solarYear,
-            self.korean_cal.solarMonth,
-            self.korean_cal.solarDay,
-        )
+        self._add_korean_new_years_eve("Vietnamese New Year's Eve")
+        self._add_korean_new_years_day("Vietnamese New Year")
+        self._add_korean_new_years_day_two("The second day of Tet Holiday")
+        self._add_korean_new_years_day_three("The third day of Tet Holiday")
+        self._add_korean_new_years_day_four("The forth day of Tet Holiday")
+        self._add_korean_new_years_day_five("The fifth day of Tet Holiday")
```

Comment:
Can we move it above the observed holidays logic?

Reply:
Yes, right, now we can move it to more natural place :)